wire/msgmixpairreq.go

```diff
@@ -203,10 +208,30 @@ func (msg *MsgMixPairReq) BtcEncode(w io.Writer, pver uint32) error {
 	return nil
 }
 
-// WriteHash serializes the message to a hasher.
+// Hash returns the message hash calculated by WriteHash.
+//
+// Hash returns an invalid or zero hash if WriteHash has not been called yet.
+//
+// This method is not safe while concurrently calling WriteHash.
+func (msg *MsgMixPairReq) Hash() chainhash.Hash {
+	return msg.hash
+}
+
+// WriteHash serializes the message to a hasher and records the sum in the
+// message's Hash field.
+//
+// The hasher's Size() must equal chainhash.HashSize, or this method will
+// panic.  This method is designed to work only with hashers returned by
+// blake256.New.
 func (msg *MsgMixPairReq) WriteHash(h hash.Hash) {
+	h.Reset()
 	writeElement(h, &msg.Signature)
 	msg.writeMessageNoSignature("", h, MixVersion)
+	sum := h.Sum(msg.hash[:0])
+	if len(sum) != len(msg.hash) {
+		s := fmt.Sprintf("hasher type %T has invalid Size() for chainhash.Hash", h)
+		panic(s)
+	}
```

Comment:
my aversion to panics is less than others, but hopefully this check is acceptable.

if there was a type safe way to enforce only blake256 hashers to be used here, that would be even better. unfortunately, this check will do nothing if you pass in a sha256 hasher, for example.

(but perhaps it might be useful that it will also allow a blake3 hash, if for whatever reason we wanted to do that one day?)

Reply:
This seems fine.  It's effectively the same as calling a func with a nil pointer when those func except non-nil inputs.

It isn't something that can be hit due to some type of unexpected decoding error or anything, rather it's only possible if the caller gives it a bogus hasher.

As an aside, and to be clear I don't expect it to change at this point given all the work and time already spent on the PR and everything built on top of it, but I still wanted to state what I mentioned elsewhere that I genuinely don't much care for the hashing, nor all of the other methods that only really exist to allow these types to be used as core data structures, being in `wire` at all.

I say that after years of dealing with `wire.MsgBlock`, `wire.MsgTx`, etc, being the core data structures used everywhere while simultaneously implementing important consensus building semantics (e.g. hashing) which makes it exceedingly difficult to separate the actual core data mechanics from the way the data happens to be serialized for sending across the wire.  I've since come to the conclusion that, in hindsight, ideally these types in `wire` should realistically have all just been limited to serialization to and from the wire as opposed to serving as the core data structures.

All of that said, I don't fault this code at all for being consistent with the existing code and there isn't another good place to them currently either as a result.